### DISCUSSION TITLE
Event handler does not read the cookies

### DISCUSSION THREAD

**Discussion by melokki** [nuxt/nuxt#16170] [Questions] [answered]
Hello everyone

I've implemented my custom authentication in my application and I am facing an issue that I can't understand how to fix it
In short, when a user authenticates, the servers responds with a token, I get get token and I save it into a cookie on the client. So far so good, my problem is that, every time I refresh the browser on a page that make a request to the api, I am getting the 401 response from the api because my event handler cannot read the cookie where the token is stored to forward it to the api server.

I don't understand how I could fix this issue. 
Do you have any thoughts/suggestions?

**Comment by TheAlexLichter** [comment, marked as answer]
**Update**: The problem is that requests can happen on the server and/or client-side in this case. When happening on the client-side, the token will be there (try out with `await useAsyncData('get-users', () => $fetch('/api/users/'), { server: false });`). When the request is happening on the server though, you have to pass the desired headers from the request along because.

Check out [the docs](https://nuxt.com/docs/getting-started/data-fetching#example-pass-cookies-from-server-side-api-calls-on-ssr-response) that show exactly your use case here ☺️

---

**Old:** Hey! Could you kindly provide a [reproduction](https://v3.nuxtjs.org/community/reporting-bugs/#create-a-minimal-reproduction)? 🙏

**Reply by melokki** [reply]
I don't know how, but I'll try

**Reply by melokki** [reply]
something like this I have in my code: https://stackblitz.com/edit/nuxt-starter-nmzd3k
on my local env, this code does not set the cookie for various reasons, could not understand why, the code is adapted from docs example.

I managed to make it work using this hack (or not, I don't know what it is):

in my event handler I placed this code:

```ts
if (!token) {
    const { token: requestToken } = event.node.req.headers;
    token = requestToken?.toString();
  }
```

and when I call it I am using this:


```ts
const { data: } = await useAsyncData(
  "get-users",
  () =>
    $fetch("/api/users", {
      headers: { token: nuxtApp.$token.value },
    })
);
```

I am sending the token from the component, because apparently it can read the cookie value

**Reply by TheAlexLichter** [reply]
The problem is that requests can happen on the server and/or client-side in this case. When happening on the client-side, the token will be there (try out with `await useAsyncData('get-users', () => $fetch('/api/users/'), { server: false });`). When the request is happening on the server though, you have to pass the desired headers from the request along because.

Check out [the docs](https://nuxt.com/docs/getting-started/data-fetching#example-pass-cookies-from-server-side-api-calls-on-ssr-response) that show exactly your use case here ☺️

**Reply by melokki** [reply]
You are my hero. Thank you very much :pray: it worked like a charm :blush:

**Reply by TheAlexLichter** [reply]
No problem at all! Glad I could help ☺️🙌🏻

**Comment by rafifn** [comment]
in pages/index.vue
```
await useAsyncData('get-users', () => $fetch('/api/users'));
```
